<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откройте-и-изучите-файл" data-toc-modified-id="Откройте-и-изучите-файл-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откройте и изучите файл</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Разбейте-данные-на-выборки" data-toc-modified-id="Разбейте-данные-на-выборки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разбейте данные на выборки</a></span></li><li><span><a href="#Исследуйте-модели" data-toc-modified-id="Исследуйте-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследуйте модели</a></span><ul class="toc-item"><li><span><a href="#Модель:-Дерево-решений" data-toc-modified-id="Модель:-Дерево-решений-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Модель: Дерево решений</a></span></li><li><span><a href="#Модель:-Случайный-лес" data-toc-modified-id="Модель:-Случайный-лес-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Модель: Случайный лес</a></span></li><li><span><a href="#Модель:-Логистическая-регрессия" data-toc-modified-id="Модель:-Логистическая-регрессия-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Модель: Логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверьте-модель-на-тестовой-выборке" data-toc-modified-id="Проверьте-модель-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверьте модель на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Модель:-Дерево-решений" data-toc-modified-id="Модель:-Дерево-решений-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Модель: Дерево решений</a></span></li><li><span><a href="#Модель:-Случайный-лес" data-toc-modified-id="Модель:-Случайный-лес-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Модель: Случайный лес</a></span></li><li><span><a href="#Модель:-Логистическая-регрессия" data-toc-modified-id="Модель:-Логистическая-регрессия-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Модель: Логистическая регрессия</a></span></li><li><span><a href="#Модель:-Случайный-лес-(валидация+обучение)" data-toc-modified-id="Модель:-Случайный-лес-(валидация+обучение)-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Модель: Случайный лес (валидация+обучение)</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#(бонус)-Проверьте-модели-на-адекватность" data-toc-modified-id="(бонус)-Проверьте-модели-на-адекватность-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>(бонус) Проверьте модели на адекватность</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

Оператор мобильной связи «Мегалайн» выяснил: многие клиенты пользуются архивными тарифами. Они хотят построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».
В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.
Постройте модель с максимально большим значением accuracy. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте accuracy на тестовой выборке самостоятельно.

### Описание данных <a class="tocSkip">

<div style="border:solid green 5px; padding: 16px; border-radius: 10px">
<br>сalls — количество звонков,
<br>minutes — суммарная длительность звонков в минутах,
<br>messages — количество sms-сообщений,
<br>mb_used — израсходованный интернет-трафик в Мб,
<br>is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).
</div>

## Откройте и изучите файл

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from tqdm import tqdm
from sklearn.dummy import DummyClassifier
pd.set_option('display.max_columns', None) #выведем все столбцы таблицы на экран

In [2]:
data = pd.read_csv("/datasets/users_behavior.csv") #Выведем таблицу на экран
data

calls  minutes  messages   mb_used  is_ultra
0      40.0   311.90      83.0  19915.42         0
1      85.0   516.75      56.0  22696.96         0
2      77.0   467.66      86.0  21060.45         0
3     106.0   745.53      81.0   8437.39         1
4      66.0   418.74       1.0  14502.75         0
...     ...      ...       ...       ...       ...
3209  122.0   910.98      20.0  35124.90         1
3210   25.0   190.36       0.0   3275.61         0
3211   97.0   634.44      70.0  13974.06         0
3212   64.0   462.32      90.0  31239.78         0
3213   80.0   566.09       6.0  29480.52         1

[3214 rows x 5 columns]

In [3]:
data.info(memory_usage='deep') #Посмотрим информацию о таблице

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [4]:
data.describe().T #Выведем стат данные о таблице

count          mean          std  min         25%        50%  \
calls     3214.0     63.038892    33.236368  0.0     40.0000     62.000   
minutes   3214.0    438.208787   234.569872  0.0    274.5750    430.600   
messages  3214.0     38.281269    36.148326  0.0      9.0000     30.000   
mb_used   3214.0  17207.673836  7570.968246  0.0  12491.9025  16943.235   
is_ultra  3214.0      0.306472     0.461100  0.0      0.0000      0.000   

                 75%       max  
calls        82.0000    244.00  
minutes     571.9275   1632.06  
messages     57.0000    224.00  
mb_used   21424.7000  49745.73  
is_ultra      1.0000      1.00

In [5]:
data.duplicated().sum() #Посмотрим количество дубликатов

0

### Вывод

По представленныым данным замечаний нет. Таблица имеет 5 столбцов и 3214 строк. Пропусков нет, дубликатов нет, тип данных менять не нужно.

## Разбейте данные на выборки

In [6]:
# Разбиваю датасет на 3 выборки:обучающую, валидационную, тестовую
df_train, df_test=train_test_split(data,test_size=0.4,random_state=1234)
df_valid,df_test=train_test_split(df_valid_test,test_size=0.5,random_state=1234)

In [7]:
#Посмотрим количество строк в выборках
print('Количество строк в обучающей выборке:', len(df_train))
print('Количество строк в валидационной выборке:', len(df_valid))
print('Количество строк в тестовой выборке:', len(df_test))

Количество строк в обучающей выборке: 1928
Количество строк в валидационной выборке: 643
Количество строк в тестовой выборке: 643


## Исследуйте модели

In [8]:
#В обучающей выборке извлечем целевой признак в переменную target_train, а остальные — в features_train
features_train=df_train.drop(['is_ultra'],axis=1)
target_train=df_train['is_ultra']

In [9]:
display(features_train.head(5)) # Признаки 
target_train.head(5) # целевой признак

calls  minutes  messages   mb_used
265    60.0   442.02      16.0  18306.45
1410   42.0   333.36      48.0  14566.34
728    43.0   298.42      12.0  18311.24
552    16.0   132.80     107.0  21893.35
3117   46.0   336.76      49.0  10710.72

265     1
1410    0
728     0
552     1
3117    0
Name: is_ultra, dtype: int64

In [10]:
#В валидационной выборке извлечем целевой признак в переменную target_valid, а остальные — в features_valid
features_valid=df_valid.drop(['is_ultra'],axis=1) # Признаки 
target_valid=df_valid['is_ultra'] # целевой признак

In [11]:
#В тестовой выборке извлечем целевой признак в переменную target_test, а остальные — в features_test
features_test=df_valid.drop(['is_ultra'],axis=1) # Признаки 
target_test=df_valid['is_ultra'] # целевой признак

### Модель: Дерево решений

In [12]:
for depth in range(1,30,2):
    model_tree=DecisionTreeClassifier(random_state=1234, max_depth=depth) # инициализируем модель DecisionTreeRegressor 
    #с параметром random_state=12345 и max_depth=depth
    model_tree.fit(features_train,target_train) # обучаем модель на тренировочной выборке
    prediction_tree=model_tree.predict(features_valid) # получаем предсказания модели на валидационной выборке
    print('max_depth:',depth,end='')
    print(' accuracy:',accuracy_score(prediction_tree,target_valid))

max_depth: 1 accuracy: 0.776049766718507
max_depth: 3 accuracy: 0.8118195956454122
max_depth: 5 accuracy: 0.8087091757387247
max_depth: 7 accuracy: 0.8320373250388803
max_depth: 9 accuracy: 0.8180404354587869
max_depth: 11 accuracy: 0.8180404354587869
max_depth: 13 accuracy: 0.8040435458786936
max_depth: 15 accuracy: 0.7744945567651633
max_depth: 17 accuracy: 0.7480559875583204
max_depth: 19 accuracy: 0.744945567651633
max_depth: 21 accuracy: 0.7620528771384136
max_depth: 23 accuracy: 0.7387247278382582
max_depth: 25 accuracy: 0.7542768273716952
max_depth: 27 accuracy: 0.7542768273716952
max_depth: 29 accuracy: 0.7402799377916018


<div style="border:solid green 5px; padding: 16px; border-radius: 10px">   
    <br> Самый лучший исход (Модель: Дерево решений):
<br>Доля правильных решений (accuracy) = 0.8320373250388803, при макисмальной глубине (max_depth) = 7 

### Модель: Случайный лес

In [13]:
best_model = None
best_result = 0
best_est = 0
best_depth = 0
for est in tqdm(range (1, 50, 2)):
    for depth in range (1, 11):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth)
        # инициализируем модель RandomForestClassifier с параметрами random_state=12345, n_estimators=est и max_depth=depth
        model.fit(features_train, target_train) # обучим модель на тренировочной выборке
        predictions_valid = model.predict(features_valid) # получим предсказания модели на валидационной выборке
        result = accuracy_score(predictions_valid,target_valid) # посчитайте значение метрики rmse на валидационной выборке
        if result > best_result:
            best_model = model
            best_result = result
            best_est = est
            best_depth = depth

print("accuracy:", best_result, "Количество деревьев:", best_est, "Максимальная глубина:", depth)

100%|██████████| 25/25 [00:13<00:00,  1.85it/s]

accuracy: 0.8429237947122862 Количество деревьев: 15 Максимальная глубина: 10


<div style="border:solid green 5px; padding: 16px; border-radius: 10px">   
    <br> Самый лучший исход (Модель: Случайный лес):
<br> Доля правильных решений (accuracy) = 0.8429237947122862, При макисмальной глубине (max_depth) = 10 и количеством деревьев = 15

### Модель: Логистическая регрессия

In [14]:
model_logistic_Reg=LogisticRegression() # инициализируем модель LogisticRegression
model_logistic_Reg.fit(features_train,target_train) # обучим модель на тренировочной выборке
predict_LogisticReg=model_logistic_Reg.predict(features_valid) # получим предсказания модели на валидационной выборке
print('accuracy:',accuracy_score(predict_LogisticReg,target_valid))

accuracy: 0.7589424572317263


### Вывод

Наилучшие показатели доли правильных решений (accuracy) для моделей:
1. Дерево решений - 0.8320373250388803
2. Случайный лес - 0.8429237947122862
3. Логистическая регрессия - 0.7589424572317263

Лучшая модель по результатам валидации - Случайный лес (0,8429)

## Проверьте модель на тестовой выборке

### Модель: Дерево решений

In [15]:
model_tree=DecisionTreeClassifier(random_state=1234, max_depth=7)
model_tree.fit(features_train,target_train)
prediction_tree=model_tree.predict(features_test)
print('Доля правильных решений (Дерево решений):',accuracy_score(prediction_tree,target_test))

Доля правильных решений (Дерево решений): 0.8320373250388803


### Модель: Случайный лес

In [16]:
model = RandomForestClassifier(random_state=12345, n_estimators=15, max_depth=10)
model.fit(features_train, target_train)
prediction_valid=model.predict(features_test)
print('Доля правильных решений (Случайный лес):',accuracy_score(prediction_valid,target_test))

Доля правильных решений (Случайный лес): 0.8429237947122862


### Модель: Логистическая регрессия

In [21]:
model_logistic_Reg=LogisticRegression(random_state=12345) 
model_logistic_Reg.fit(features_train,target_train)
predict_LogisticReg=model_logistic_Reg.predict(features_test) 
print('Доля правильных решений (Логистическая регрессия):',accuracy_score(predict_LogisticReg,target_test))

Доля правильных решений (Логистическая регрессия): 0.7589424572317263


### Модель: Случайный лес (валидация+обучение)

In [18]:
#Объединим признаки из валидации и обучения
features_df = features_train.append(features_valid)
target_df = target_train.append(target_valid)

In [19]:
model = RandomForestClassifier(random_state=12345, n_estimators=15, max_depth=10)
model.fit(features_df, target_df)
prediction_valid=model.predict(features_test)
print('Доля правильных решений (Случайный лес):',accuracy_score(prediction_valid,target_test))

Доля правильных решений (Случайный лес): 0.9066874027993779


### Вывод

При проверке по тестовой выборке, метрики качества тестовой выборки=метрикам качества валидационной модели:
1. Дерево решений - 0.8320373250388803
2. Случайный лес - 0.8429237947122862
3. Логистическая регрессия - 0.7589424572317263

## (бонус) Проверьте модели на адекватность

In [20]:
#Построим простейшую (dummy) модель, которая всегда предсказывает наиболее часто встречающийся класс.
dummy_clf = DummyClassifier(strategy="most_frequent", random_state=0)
dummy_clf.fit(features_train, target_train)
print('Доля правильных решений (Простейшая модель):',dummy_clf.score(features_train, target_train))

Доля правильных решений (Простейшая модель): 0.6877593360995851


### Вывод

Поскольку показатель доли правльных решений у базовой (простейшей модели) значительно меньше (0,68), мы можем говорить об адекватности наших моделей.

## Общий вывод

<div style="border:solid green 5px; padding: 20px">
    
**<font color="green">Общий анализ данных датасета</font>**
<br>По представленныым данным замечаний нет. Таблица имеет 5 столбцов и 3214 строк. Пропусков нет, дубликатов нет, тип данных менять не нужно.
    
**<font color="green">Исследование моделей</font>**   
<br>Наилучшие показатели доли правильных решений (accuracy) для моделей:
<br>1. Дерево решений - 0.8320373250388803
<br>2. Случайный лес - 0.8429237947122862
<br>3. Логистическая регрессия - 0.7589424572317263
  
**<font color="green">Проверка моделей на тестовой выборке</font>**     
<br>При проверке по тестовой выборке, получили следующие показатели доли правильных решений (accuracy) для моделей:
<br>1. Дерево решений - 0.7402799377916018
<br>2. Случайный лес - 0.8398133748055988
<br>3. Логистическая регрессия - 0.7589424572317263    
    
**<font color="green">Проверка моделей на адекватность</font>**  
<br>Поскольку показатель доли правльных решений у базовой (простейшей модели) значительно меньше (0,68), мы можем говорить об адекватности наших моделей.

**<font color="green">Итог</font>** 
<br>Наилучшая модель - Случайный лес со следующими показателями доли правильных решений (accuracy):
<br>0.8429237947122862 - при проверке на валидационной выборке
<br>0.8398133748055988 - при проверке на тестовой выборке 